### PR TITLE
Adding parallel support for the LLVM backend.

### PR DESCRIPTION
**Pull request by zheng-xq**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#53243 Adding parallel support for the LLVM backend.**

Differential Revision: [D26906509](https://our.internmc.facebook.com/intern/diff/D26906509)

### CODE REVIEW

File: torch/csrc/jit/tensorexpr/llvm_codegen.cpp [hunk 1]
```diff
@@ -229,6 +242,39 @@ class LLVMCodeGenImpl : public IRVisitor {
 
   void optimize(llvm::Module& M);
 };
+
+// TODO: more extensive testing.
+// TODO: benchmarking.
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/llvm_codegen.cpp:247]: There are some benchmarks in the PR, do you plan to add something more? If so, it would be helpful to elaborate in the comment (and similarly in the previous comment about more tests).

File: test/cpp/tensorexpr/test_llvm.cpp [hunk 2]
```diff
@@ -1610,6 +1610,133 @@ TEST(LLVM, RFactorVectorizedReduction) {
...
+    ASSERT_EQ(loop_list.size(), loop_count);
+    for (int i = 0; i < loop_count; i++) {
+      if (test_cfg & (1 << i)) {
+        loop_list[i]->set_parallel();
+      }
+    }
+    Placeholder t1_buf(BufHandle(t1->buf()));
+    Placeholder t2_buf(BufHandle(t2->buf()));
+    Placeholder t3_buf(BufHandle(t3->buf()));
+    Placeholder t4_buf(BufHandle(t4->buf()));
```
Comment (ZolotukhinM) [inline test/cpp/tensorexpr/test_llvm.cpp:1715]: You do not need to create placeholders for Tensors, they are only needed for external buffers, for which we don't have `Compute`.

File: test/cpp/tensorexpr/test_llvm.cpp [hunk 3]
```diff
@@ -1610,6 +1610,133 @@ TEST(LLVM, RFactorVectorizedReduction) {
...
+    loop_nest.prepareForCodegen();
+    Stmt* stmt = loop_nest.root_stmt();
+    LLVMCodeGen cg(stmt, {t4_buf});
```
Comment (ZolotukhinM) [inline test/cpp/tensorexpr/test_llvm.cpp:1718]: You could pass `t4` directly here:
```
LLVMCodeGen cg(stmt, {t4});
```
I noticed that there are other tests doing a similar thing, but I believe that's accidental or some leftovers.

File: test/cpp/tensorexpr/test_llvm.cpp [hunk 4]
```diff
@@ -1610,6 +1610,133 @@ TEST(LLVM, RFactorVectorizedReduction) {
...
+    if (test_cfg & 0x1) {
+      m->set_parallel();
+    }
+    if (test_cfg & 0x2) {
+      n->set_parallel();
+    }
+    loop_nest.prepareForCodegen();
+    Stmt* stmt = loop_nest.root_stmt();
+    Placeholder f_buf(BufHandle(tensor->buf()));
+    LLVMCodeGen cg(stmt, {f_buf});
```
Comment (ZolotukhinM) [inline test/cpp/tensorexpr/test_llvm.cpp:1638]: Placeholder is not needed here, you could pass `tensor` directly to the codegen constructor:
```
LLVMCodeGen cg(stmt, {tensor});
```

File: torch/csrc/jit/tensorexpr/llvm_codegen.cpp [hunk 5]
```diff
@@ -447,6 +494,7 @@ class LLVMIntrinsicsExpander : public GenericIntrinsicsExpander {
 
 void LLVMCodeGenImpl::emitKernel(
     Stmt* stmt,
+
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/llvm_codegen.cpp:497]: Nit: unnecessary whitespace change.

File: torch/csrc/jit/tensorexpr/llvm_codegen.cpp [hunk 6]
```diff
@@ -1109,7 +1180,133 @@ void LLVMCodeGenImpl::visit(const Load* v) {
...
+  // Corresponding Value* that will be used in the new body closure.
+  std::vector<llvm::Value*> body_closure_args;
+
+  // Identify the Var* used in the body, and generated outside.
+  VarFinder var_finder;
+  v->body()->accept(&var_finder);
+  const auto& vars = var_finder.vars();
+  for (auto& var : vars) {
+    llvm::Value* value = varToValue(var);
+    if (value) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/llvm_codegen.cpp:1246]: Nit: can do in one line :)

**Review by ZolotukhinM — APPROVED** [review]
This is awesome! I have just a couple of small nits, but since the code is quite complicated (because it solves quite a complicated problem!) I'd recommend to wait for someone else to take one more look at it.

**Comment by zheng-xq** [comment]
All comments addressed. PTAL. Thanks!

**Review by navahgar — APPROVED** [review]
LGTM

File: torch/csrc/jit/tensorexpr/llvm_codegen.cpp [hunk 7]
```diff
@@ -139,6 +170,8 @@ class LLVMCodeGenImpl : public IRVisitor {
   AT_FORALL_SCALAR_TYPES_AND2(Bool, Half, LLVM_TYPE_DECLARE);
 #undef LLVM_TYPE_DECLARE
   llvm::Type* Int8PtrTy_;
+  llvm::Type* VoidTy_;
+  llvm::Type* VoidPtrTy_;
```
Comment (bertmaher) [inline torch/csrc/jit/tensorexpr/llvm_codegen.cpp:174]: Looks like this (`VoidPtrTy_`) is unused?

File: torch/csrc/jit/tensorexpr/llvm_codegen.cpp [hunk 8]
```diff
@@ -229,6 +270,37 @@ class LLVMCodeGenImpl : public IRVisitor {
...
+      USE_TRIGGER(llvm_codegen_parallel_dispatched);
+    } else {
+      // the current thread handles the last one.
+      callee(index, packed_data);
+      USE_TRIGGER(llvm_codegen_parallel_dispatched);
+      std::unique_lock<std::mutex> l(m);
+      remaining_tasks--;
+      while (remaining_tasks != 0) {
+        cv.wait(l);
+      }
```
Comment (bertmaher) [inline torch/csrc/jit/tensorexpr/llvm_codegen.cpp:299]: I usually like to use the predicate version of `wait` e.g. `cv.wait(l, [&] { remaining == 0 })` but this is fine too

File: torch/csrc/jit/tensorexpr/llvm_jit.h [hunk 9]
```diff
@@ -16,6 +16,13 @@ namespace torch {
 namespace jit {
 namespace tensorexpr {
 
+void DispatchParallel(int8_t* func, int start, int stop, int8_t* packed_data);
+typedef void (*TypeDispatchParallel)(
+    int8_t* func,
+    int start,
+    int stop,
+    int8_t* packed_data);
+
```
Comment (bertmaher) [inline torch/csrc/jit/tensorexpr/llvm_jit.h:25]: TypeDispatchParallel looks unused, remove if so.

File: torch/csrc/jit/tensorexpr/llvm_codegen.cpp [hunk 10]
```diff
@@ -1136,6 +1357,7 @@ void LLVMCodeGenImpl::visit(const For* v) {
   auto exit = llvm::BasicBlock::Create(getContext(), "exit", fn_);
 
   // Create the stop condition.
+  irb_.SetInsertPoint(condBlock);
```
Comment (bertmaher) [inline torch/csrc/jit/tensorexpr/llvm_codegen.cpp:1360]: Is this needed?  I'm a little surprised since we just do `irb_.SetInsertPoint(condBlock);` a few lines ago (1344) and I don't see anything that would invalidate that...

**Review by bertmaher — APPROVED** [review]
Very cool!

**Review by bertmaher — APPROVED** [review]

**Comment by ZolotukhinM** [comment]
Looks like a number of approves and reaccs record! 👍
